static/src/javascripts/projects/commercial/modules/article-body-adverts.js

```diff
@@ -0,0 +1,209 @@
+
+        return Promise.all(slots).then(() => slots.length);
+    };
+
+    return spaceFiller.fillSpace(rules, insertInlineAds, {
+        waitForImages: true,
+        waitForLinks: true,
+        waitForInteractives: true,
+    });
+};
+
+const getRules = (): Object => {
+    let prevSlot;
+    return {
+        bodySelector: '.js-article__body',
+        slotSelector: ' > p',
+        minAbove: detect.isBreakpoint({
+            max: 'tablet',
+        })
+            ? 300
+            : 700,
+        minBelow: 300,
+        selectors: {
+            ' > h2': {
+                minAbove: detect.getBreakpoint() === 'mobile' ? 100 : 0,
+                minBelow: 250,
+            },
+            ' .ad-slot': {
+                minAbove: 500,
+                minBelow: 500,
+            },
+            ' > :not(p):not(h2):not(.ad-slot)': {
+                minAbove: 35,
+                minBelow: 400,
+            },
+        },
+        filter(slot) {
+            if (
+                !prevSlot ||
+                Math.abs(slot.top - prevSlot.top) - adSizes.mpu.height >=
+                    this.selectors[' .ad-slot'].minBelow
+            ) {
+                prevSlot = slot;
+                return true;
+            }
+            return false;
+        },
+    };
+};
+
+const getAltRules = () => {
+    const altRules = getRules();
+    altRules.selectors = {
+        ' .ad-slot': {
+            minAbove: 500,
+            minBelow: 500,
+        },
+    };
+    return altRules;
+};
+
+const getLongArticleRules = () => {
+    const longArticleRules = isOffsetingAds ? getAltRules() : getRules();
+    const viewportHeight = detect.getViewport().height;
+    longArticleRules.selectors[' .ad-slot'].minAbove = viewportHeight;
+    longArticleRules.selectors[' .ad-slot'].minBelow = viewportHeight;
+    return longArticleRules;
+};
+
+const addInlineAds = () =>
+    addArticleAds(
+        2,
+        isOffsetingAds ? getAltRules() : getRules()
+    ).then(countAdded => {
+        if (countAdded === 2) {
+            return addArticleAds(8, getLongArticleRules()).then(
+                innerCountAdded => 2 + innerCountAdded
+            );
+        }
+        return countAdded;
+    });
+
+const getInlineMerchRules = () => {
+    const inlineMerchRules = getRules();
+    inlineMerchRules.minAbove = 300;
+    inlineMerchRules.selectors[' > h2'].minAbove = 100;
+    inlineMerchRules.selectors[
+        ' > :not(p):not(h2):not(.ad-slot)'
+    ].minAbove = 200;
+    return inlineMerchRules;
+};
+
+const addInlineMerchAd = () =>
+    spaceFiller.fillSpace(
+        getInlineMerchRules(),
+        paras => insertAdAtPara(paras[0], 'im', 'im').then(() => 1),
+        {
+            waitForImages: true,
+            waitForLinks: true,
+            waitForInteractives: true,
+        }
+    );
+
+const waitForMerch = (countAdded: number) =>
+    countAdded === 1 ? trackAdRender('dfp-ad--im') : Promise.resolve();
+
+const init = () => {
```

Comment:
I think a signature would be useful here. The commercial bootstrap relies on that single contract that each coroutine is a `() => Promise<T>` and so we could verify that at compile-time.